src/iosrv/sched.c

```diff
@@ -1359,10 +1384,106 @@ sched_pop_one(struct sched_data *data, ABT_pool pool, int pool_idx)
 	return unit;
 }
 
+#define SCHED_IDLE_THRESH	8000UL	/* msecs */
+
+/*
+ * Try to relax CPU for a short period when the xstream is idle. The relaxing
+ * period can't be too long, otherwise, potential external events like:
+ * incoming network requests, new ULTs created by other xstream (from the
+ * collective call or offloading call) could be delayed too much.
+ *
+ * There are also some periodical internal events from BIO, like hotplug
+ * poller, health/io stats collecting, blobstore state transition, etc. It's
+ * not easy to accurately predict the next occurrence of those events.
+ */
+static void
+sched_try_relax(struct dss_xstream *dx, ABT_pool *pools, uint32_t running)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	unsigned int		 sleep_time = sched_relax_intvl;
+	size_t			 blocked;
+	int			 ret;
+
+	dx->dx_timeout = 0;
+
+	if (info->si_stop)
+		return;
+
+	/*
+	 * There are running ULTs in current schedule cycle.
+	 *
+	 * NB. DRPC listener ULT is currently always running (it waits on
+	 * drpc_progress(), so the DRPC listener xstream will never sleep
+	 * in this function.
+	 */
+	if (running != 0)
+		return;
+
+	/* There are queued requests to be processed */
+	if (info->si_req_cnt != 0)
+		return;
+
+	ret = ABT_pool_get_total_size(pools[DSS_POOL_GENERIC], &blocked);
+	if (ret != ABT_SUCCESS) {
+		D_ERROR("XS(%d) get ABT pool(%d) total size error: %d\n",
+			dx->dx_xs_id, DSS_POOL_GENERIC, ret);
+		return;
+	}
+	D_ASSERTF(info->si_sleep_cnt + info->si_wait_cnt <= blocked,
+		  "sleep:%d + wait:%d > blocked:%zd\n",
+		  info->si_sleep_cnt, info->si_wait_cnt, blocked);
+
+	/*
+	 * Only start relaxing when all blocked ULTs are either sleeping
+	 * ULT or long wait ULT.
+	 */
+	if (blocked > info->si_sleep_cnt + info->si_wait_cnt)
+		return;
+
+	/*
+	 * System is currently idle, but we only start relaxing when there is
+	 * no external events for a short period of SCHED_IDLE_THRESH.
+	 */
+	D_ASSERT(info->si_cur_ts >= info->si_stats.ss_busy_ts);
+	if (info->si_cur_ts - info->si_stats.ss_busy_ts < SCHED_IDLE_THRESH)
```

Comment:
Why are we not maintaining ss_busy_ts in sched_run? Could we determine that "ah, there's no more runnable or immediate task now" in sched_run?

Reply:
busy timestamp is bumped once any external event arrives, so that scheduler won't relax unless there is no external event arrives in certain period.

Scheduler knows when there is no runnable and immediate tasks, but it doesn't mean we can enter relax mode immediately, we need to keep busy polling until no external event comes in for certain period of time.